src/deploy/mod.rs

```diff
@@ -1,82 +1,95 @@
 use std::io::Write;
 use std::{ops::DerefMut, sync::Arc, time::SystemTime};
 use tokio::sync::Semaphore;
 
-#[derive(serde::Serialize, serde::Deserialize)]
-pub struct CalculateRequest {
-    pub beatmap_id: i32,
-    pub beatmap_md5: String,
-    pub mode: i32,
+#[derive(Clone, sqlx::FromRow)]
+struct LightweightScore {
+    pub id: i64,
     pub mods: i32,
     pub max_combo: i32,
-    pub count_300: i32,
-    pub count_100: i32,
-    pub count_50: i32,
-    pub miss_count: i32,
-}
-
-#[derive(serde::Serialize, serde::Deserialize)]
-pub struct CalculateResponse {
-    pub stars: f32,
+    pub play_mode: i32,
+    pub beatmap_id: i32,
     pub pp: f32,
+    pub accuracy: f32,
+
+    #[sqlx(rename = "misses_count")]
+    pub count_misses: i32,
 }
 
 fn round(x: f32, decimals: u32) -> f32 {
     let y = 10i32.pow(decimals) as f32;
     (x * y).round() / y
 }
 
-const RX: i32 = 1 << 7;
-const AP: i32 = 1 << 13;
-
-async fn calculate_special_pp(
-    request: &CalculateRequest,
-    context: Arc<Context>,
-) -> anyhow::Result<CalculateResponse> {
-    let beatmap_bytes =
-        usecases::beatmaps::fetch_beatmap_osu_file(request.beatmap_id, context.clone()).await?;
-    let beatmap = Beatmap::from_bytes(&beatmap_bytes)?;
+const MAX_CONCURRENT_TASKS: usize = 100;
+const BATCH_SIZE: u32 = 1000;
 
-    let result = akatsuki_pp_rs::osu_2019::OsuPP::new(&beatmap)
-        .mods(request.mods as u32)
-        .combo(request.max_combo as u32)
-        .misses(request.miss_count as u32)
-        .n300(request.count_300 as u32)
-        .n100(request.count_100 as u32)
-        .n50(request.count_50 as u32)
-        .calculate();
+fn group_scores_by_mods(vec: Vec<LightweightScore>) -> HashMap<i32, Vec<LightweightScore>> {
+    let mut grouped_map = HashMap::new();
 
-    let mut pp = round(result.pp as f32, 2);
-    if pp.is_infinite() || pp.is_nan() {
-        pp = 0.0;
+    for item in vec {
+        grouped_map
+            .entry(item.mods)
+            .or_insert_with(Vec::new)
+            .push(item);
     }
 
-    let mut stars = round(result.difficulty.stars as f32, 2);
-    if stars.is_infinite() || stars.is_nan() {
-        stars = 0.0;
+    grouped_map
+}
+
+async fn recalculate_relax_scores(
+    scores: Vec<LightweightScore>,
+    mods: i32,
+    scores_table: &str,
+    beatmap: &Beatmap,
+    ctx: Arc<Context>,
+) -> anyhow::Result<()> {
+    let difficulty_attributes =
+        akatsuki_pp_rs::osu_2019::stars::stars(&beatmap, (mods as u32).into());
+
+    for score in scores {
+        let result =
+            akatsuki_pp_rs::osu_2019::OsuPP::from_attributes(difficulty_attributes.clone())
+                .mods(score.mods as u32)
+                .combo(score.max_combo as u32)
+                .misses(score.count_misses as u32)
+                .accuracy(score.accuracy)
+                .calculate();
+
+        let mut pp = round(result.pp as f32, 2);
+        if pp.is_infinite() || pp.is_nan() {
+            pp = 0.0;
```

Comment:
should we log these cases?

Reply:
not frequent afaik, but it's not benched so can't give a real answer

Reply:
Maybe worth a trial run on a smaller gamemode on prod?